src/fides/api/models/privacy_request.py

```diff
@@ -1750,15 +1772,79 @@ def get_pending_downstream_tasks(self, db: Session) -> Query:
+        ) and not self.request_task_running(should_log)
+
+    def upstream_tasks_complete(self, db: Session, should_log: bool = False) -> bool:
+        """Determines if all of the upstream tasks of the current task are complete"""
         upstream_tasks: Query = db.query(RequestTask).filter(
             RequestTask.privacy_request_id == self.privacy_request_id,
             RequestTask.collection_address.in_(self.upstream_tasks or []),
             RequestTask.action_type == self.action_type,
         )
-
-        return all(
+        tasks_complete: bool = all(
             upstream_task.status in COMPLETED_EXECUTION_LOG_STATUSES
             for upstream_task in upstream_tasks
+        ) and upstream_tasks.count() == len(self.upstream_tasks or [])
+
+        if not tasks_complete and should_log:
+            logger.debug(
+                "Upstream tasks incomplete for {} task {}. Privacy Request: {}, Request Task {}.",
+                self.action_type.value,
+                self.collection_address,
+                self.privacy_request_id,
+                self.id,
+            )
+
+        return tasks_complete
+
+    def upstream_tasks_objects(self, db: Session) -> Query:
+        """Returns Request Task objects that are upstream of the current Request Task"""
+        upstream_tasks: Query = db.query(RequestTask).filter(
+            RequestTask.privacy_request_id == self.privacy_request_id,
+            RequestTask.collection_address.in_(self.upstream_tasks or []),
+            RequestTask.action_type == self.action_type,
         )
+        return upstream_tasks
+
+    def request_task_running(self, should_log: bool = False) -> bool:
+        """Returns a rough measure if the Request Task is already running -
+
+        This is only applicable if you are running workers and
+        CONFIG.execution.task_always_eager=False. This is just an extra check to reduce possible
+        over-scheduling, but it is also okay if the same node runs multiple times.
+        """
+        celery_task_id: Optional[str] = self.get_cached_task_id()
+        if not celery_task_id:
+            return False
+
+        if should_log:
+            logger.debug(
+                "Celery Task ID {} found for {} task {}. Privacy Request: {}, Request Task {}.",
+                celery_task_id,
+                self.action_type.value,
+                self.collection_address,
+                self.privacy_request_id,
+                self.id,
+            )
+
+        task_in_flight: bool = celery_tasks_in_flight([celery_task_id])
+
+        if task_in_flight and should_log:
+            logger.debug(
+                "Celery Task {} already processing for {} task {}. Privacy Request: {}, Request Task {}.",
+                celery_task_id,
+                self.action_type.value,
+                self.collection_address,
+                self.privacy_request_id,
+                self.id,
+            )
+
+        return task_in_flight
```

Comment:
I added a light check to reduce some of the over-scheduling if two nodes running in parallel finish about the same time, realize that the other is finished, and go to queue their shared downstream node about the same time.  It is hard to hit this state in testing with workers, but possible.

Reply:
Thank you Adrian!